### DISCUSSION TITLE
Add checkbox to use cdn for engine wasm

### DISCUSSION THREAD

**Discussion by dwidge** [godotengine/godot-proposals#8619] [Platforms]
Can we add a checkbox to the Web export to use a public cdn hosted index.wasm file?

The cdn could have the index.wasm debug/release for each version of Godot.
Could deploy new index.wasm versions to the cdn here: https://github.com/godotengine/godot/blob/master/.github/workflows/web_builds.yml
The correct cdn url will be selected automatically based on Godot version and debug/release.

Benefits
- Decrease game bundle size
- Decrease load times
- Browser can cache it and share with many games on same domain
- Game can be updated without redownloading engine
- The cdn will serve it gzipped at the edge
- Less storage/bandwidth cost for game host

Example bundle size
- With index.wasm: 34.6MB / 7.6MB zip
- No index.wasm: 1MB / 350KB zip

Some people already host their own shared engine wasm:
https://www.reddit.com/r/godot/comments/pjuqsr/html5_could_godots_wasm_file_be_hosted_in_a/
But a checkbox could be easier.

**Comment by Calinou** [comment]
This is an interesting idea, but CDNs for static resources have started to fall out of fashion for a few reasons lately. They're not *always* faster[^1] and there's been increased privacy concerns lately. Security isn't too much of a concern nowadays if you use [Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity), but long-term availability still is. A local file will always be available on the same server in the future, while free public CDNs don't make any guarantees about availability.

[^1]: There's been numerous tests in the wild where websites hosted purely on a single server turned out to be faster to load than the same website using static assets hosted on a CDN.

It makes sense to use a CDN if you host your project on a platform where you pay for bandwidth, still. That said, we don't have an official platform we can provide CDN builds on.

> - Decrease game bundle size

The downloaded file size will be exactly the same, regardless of whether you use a CDN or not.

> - Decrease load times

This is only the case if the file can be reused from another project on the same domain name. I don't expect this to be very common on most websites. I believe dedicated platforms like gotm.io already use a shared WASM runtime.

> - The cdn will serve it gzipped at the edge

It's possible to enable gzip/Brotli compression on your own web server, but this is unfortunately not done by all hosts out there for WASM files (let alone PCK). This could be an argument to use CDNs to circumvent this issue on hosts that don't use gzip/Brotli compression… but only if they don't use a CSP that prevents loading external resources.

> - Game can be updated without redownloading engine

If you don't change the engine version between two updates, the `.wasm` file will retain its original checksum. That said, whether the web server will always reserve it depends on its caching configuration.

**Reply by dwidge** [reply]
Ok it won't be that useful. Thank you for all the details 😃
